app/controllers/api/songs_controller.rb

```diff
@@ -0,0 +1,102 @@
+class API::SongsController < API::APIController
+
+  before_action :require_edit_privileges, only: [:create, :update]
+  before_action :require_delete_privileges, only: [:destroy]
+
+  def show
+    song = Song.find_by(id: params[:id])
+    if song
+      case params[:format]
+      when 'numbers'
+        Formatter.format_song_nashville(song)
+      when 'no_chords'
+        Formatter.format_song_no_chords(song)
+      else
+        Transposer.transpose_song(song, params[:key]) if params[:key].present?
+      end
+
+      Song.increment_counter(:view_count, song.id, touch: false)
+      render json: song
+    else
+      head :not_found
+    end
+  end
+
+  # Query params are single letter to save space, since HTTP urls
+  def index
+    songs = Song
+
+    # perform searching and filtering
+    # pg_search requires search_by_keywords to be to run first before anything else
+    songs = songs.search_by_keywords(params[:query]) if params[:query].present?
+    songs = songs.where(key: params[:key]) if params[:key].present?
+    songs = songs.where(tempo: params[:tempo]) if params[:tempo].present?
+    songs = songs.select('id, artist, tempo, key, name, chord_sheet, spotify_uri')
+
+    # reorder
+    songs =
+      case params[:sort_by]
+      when 'created_at'
+        songs.reorder(created_at: :desc)
+      when 'views'
+        songs.reorder(view_count: :desc)
+      else
+        # does nothing if search_by_keywords was run, in which case songs are already ordered by relevance
+        songs.order(name: :asc)
+      end
+
+    matching_songs_count = songs.size
+
+    # paginate
+    page_num = params[:page_num]&.to_i || 1
```

Comment:
Joseph, is the `&.` idiomatic Ruby? I hardly see it get used when I study OSS Rails projects to learn.

Reply:
yes this is totally fine use case of safe navigator